include/RED4ext/SharedPtr.hpp

```diff
@@ -0,0 +1,377 @@
+
+template<typename T>
+class SharedPtrBase
+{
+public:
+    SharedPtrBase(const SharedPtrBase&) = delete;
+
+    ~SharedPtrBase() = default;
+
+    SharedPtrBase& operator=(const SharedPtrBase&) = delete;
+
+    [[nodiscard]] uint32_t GetUseCount() const noexcept
+    {
+        return refCount ? refCount->strongRefs : 0;
+    }
+
+    T* instance;
+    RefCnt* refCount;
+
+protected:
+    constexpr SharedPtrBase() noexcept
+        : instance(nullptr)
+        , refCount(nullptr)
+    {
+    }
+
+    void DoSwap(SharedPtrBase& aOther) noexcept
+    {
+        std::swap(instance, aOther.instance);
+        std::swap(refCount, aOther.refCount);
+    }
+
+    void CopyConstructFrom(const SharedPtrBase& aOther) noexcept
+    {
+        aOther.IncRef();
+        instance = aOther.instance;
+        refCount = aOther.refCount;
+    }
+
+    bool TryCopyConstructFromWeak(const SharedPtrBase& aOther) noexcept
+    {
+        if (aOther.IncRefIfNotZero())
+        {
+            instance = aOther.instance;
+            refCount = aOther.refCount;
+            return true;
+        }
+
+        return false;
+    }
+
+    void MoveConstructFrom(SharedPtrBase&& aOther) noexcept
+    {
+        instance = aOther.instance;
+        refCount = aOther.refCount;
+        aOther.instance = nullptr;
+        aOther.refCount = nullptr;
+    }
+
+    void CopyConstructWeakFrom(const SharedPtrBase& aOther) noexcept
+    {
+        instance = aOther.instance;
+        refCount = aOther.refCount;
+        IncWeakRef();
+    }
+
+    void IncRef() const noexcept
+    {
+        if (refCount)
+        {
+            refCount->IncRef();
+        }
+    }
+
+    [[nodiscard]] bool IncRefIfNotZero() const noexcept
+    {
+        if (refCount)
+        {
+            return refCount->IncRefIfNotZero();
+        }
+
+        return false;
+    }
+
+    [[nodiscard]] bool DecRef() noexcept
+    {
+        if (refCount && refCount->DecRef())
+        {
+            DecWeakRef();
+            return true;
+        }
+
+        return false;
+    }
+
+    void Destroy() noexcept
+    {
+        Memory::Delete(instance);
+
+        instance = nullptr;
+        refCount = nullptr;
+    }
+
+    void IncWeakRef() const noexcept
+    {
+        if (refCount)
+        {
+            refCount->IncWeakRef();
+        }
+    }
+
+    // This is here and not in RefCnt because the function we use expects a pointer to a SharedPtr.
+    void DecWeakRef()
+    {
+        if (refCount)
+        {
+            // This function also deallocates RefCnt when weakRefs reaches 0.
+            RelocFunc<void (*)(SharedPtrBase*)> decWeakRefFn(Addresses::Handle_DecWeakRef);
+            decWeakRefFn(this);
+        }
+    }
+};
+RED4EXT_ASSERT_SIZE(SharedPtrBase<void>, 0x10);
+RED4EXT_ASSERT_OFFSET(SharedPtrBase<void>, instance, 0x00);
+RED4EXT_ASSERT_OFFSET(SharedPtrBase<void>, refCount, 0x08);
+
+template<typename T>
+class SharedPtrWithAccess : public SharedPtrBase<T>
```

Comment:
Wouldn't it be better to implement these directly on the `SharedPtr` class?

Reply:
And copy-paste to `Handle`?

Reply:
Oh, nvm then. I missed that part.